google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,512 @@
+    future<void> retry_future;
+    std::lock_guard<std::mutex> g{mu_};
+    switch (stream_state_) {
+      case State::kShutdown:
+        return make_ready_future();
+      case State::kRetrying:
+        stream_state_ = State::kShutdown;
+        retry_future = retry_promise_->get_future();
+        mu_.unlock();
+        return retry_future.then([this](future<void>) {
+          std::lock_guard<std::mutex> g{mu_};
+          CompleteStreamLockHeldNotRelocked(Status());
+        });
+      case State::kInitialized:
+        stream_state_ = State::kShutdown;
+        future<void> to_finish_future = make_ready_future();
+        if (in_progress_read_) {
+          auto read_future =
+              std::make_shared<future<void>>(in_progress_read_->get_future());
+          to_finish_future = to_finish_future.then(
+              [read_future](future<void>) { return std::move(*read_future); });
+        }
+        if (in_progress_write_) {
+          auto write_future =
+              std::make_shared<future<void>>(in_progress_write_->get_future());
+          to_finish_future =
+              to_finish_future.then([write_future](future<void>) {
+                return std::move(*write_future);
+              });
+        }
+        std::shared_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+            stream = std::move(stream_);
+        CompleteStreamLockHeldNotRelocked(Status());
+        return to_finish_future.then([stream](future<void>) {
+          return stream->Finish().then(
+              [](future<Status>) { return make_ready_future(); });
+        });
+    }
+  }
+
+ private:
+  enum class State { kRetrying, kInitialized, kShutdown };
+
+  void ReadWriteRetryFailedStream() {
+    promise<void> root;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (stream_state_ != State::kInitialized) return;
+
+      stream_state_ = State::kRetrying;
+      retry_promise_.emplace();
+
+      // Assuming that a `Read` fails:
+      // If an outstanding operation is present, we can't enter the retry
+      // loop, so we defer it until the outstanding `Write` finishes at
+      // which point we can enter the retry loop. Since we will return
+      // `reinit_done_`, we guarantee that another operation of the same type
+      // is not called while we're waiting for the outstanding operation to
+      // finish and the retry loop to finish afterward.
+
+      future<void> root_future = root.get_future();
+
+      if (in_progress_read_.has_value()) {
+        assert(!in_progress_write_.has_value());
+        auto in_progress =
+            std::make_shared<future<void>>(in_progress_read_->get_future());
+        root_future = root_future.then(
+            [in_progress](future<void>) { return std::move(*in_progress); });
+      } else if (in_progress_write_.has_value()) {
+        auto in_progress =
+            std::make_shared<future<void>>(in_progress_write_->get_future());
+        root_future = root_future.then(
```

Comment:
in pubsublite/internal/futures.h, you should add a helper:

```
template <class T>
std::function<future<void>(future<void>)> ChainFuture(future<T> f) {
  auto to_chain = std::make_shared<future<void>>(std::move(f));
  return [=](future<void>){ return std::move(*to_chain); };
}
```

Then you can use that here.